lib/veres-one-did-doc.js

```diff
@@ -0,0 +1,452 @@
+
+    // verify against the key
+    if(!keyPair.verifyFingerprint(fingerprint)) {
+      throw new Error(`Invalid DID - fingerprint does not verify against key`);
+    }
+  }
+
+  /**
+   * Starts observing changes made to the DID Document, in order to create a
+   * diff patch to send to the ledger. Used for updating the doc.
+   */
+  observe() {
+    if(this.observer) {
+      this.unobserve();
+    }
+    this.observer = jsonpatch.observe(this.doc);
+  }
+
+  /**
+   * Stops observing for changes.
+   */
+  unobserve() {
+    if(!this.observer) {
+      throw new Error('Not observing changes');
+    }
+    jsonpatch.unobserve(this.doc, this.observer);
+    this.observer = null;
+  }
+
+  /**
+   * Stops observing for changes, and returns a changeset document (based on
+   * JSON Patch), for sending updates to ledger.
+   *
+   * @returns {object}
+   */
+  commit() {
+    if(!this.observer) {
+      throw new Error('Not observing changes');
+    }
+    const patch = jsonpatch.generate(this.observer);
+
+    this.unobserve();
+
+    return {
+      '@context': constants.VERES_ONE_V1_CONTEXT,
+      target: this.id,
+      sequence: this.meta.sequence,
+      patch
+    };
+  }
+
+  /**
+   * Composes and returns a service id for a service name.
+   *
+   * @param serviceName {string}
+   *
+   * @returns {string} Service id
+   */
+  serviceIdFor(serviceName) {
+    if(!serviceName) {
+      throw new Error('Invalid service name');
+    }
+
+    return this.id + ';service=' + encodeURIComponent(serviceName);
+  }
+
+  /**
+   * Finds a service endpoint in this did doc, given an id or a name.
+   *
+   * @param [id] {string}
+   * @param [name] {string}
+   *
+   * @returns {object}
+   */
+  findService({id, name}) {
+    const jsonld = this.injector.use('jsonld');
+
+    const serviceId = id || this.serviceIdFor(name);
+
+    return jsonld
+      .getValues(this.doc, 'service')
+      .find(service => service.id === serviceId);
+  }
+
+  /**
+   * Tests whether this did doc has a service endpoint (by name or id).
+   * One of `id` or `name` is required.
+   *
+   * @param [id] {string}
+   * @param [name] {string}
+   *
+   * @returns {boolean}
+   */
+  hasService({id, name}) {
+    return !!this.findService({id, name});
+  }
+
+  /**
+   * Adds a service endpoint to this did doc.
+   * One of `id` or `name` is required.
+   *
+   * @param [id] {string}
+   * @param [name] {string}
+   * @param type {string} Endpoint type (e.g. 'AgentService')
+   * @param endpoint {string} Endpoint uri (e.g. 'https://agent.example.com')
+   * @param [options] {object} Any additional properties of endpoint
+   */
+  addService({id, name, type, endpoint, ...options}) {
+    const jsonld = this.injector.use('jsonld');
+
+    const serviceId = id || this.serviceIdFor(name);
+
+    if(!type) {
+      throw new Error('Service endpoint type is required');
+    }
+    if(!endpoint) {
+      throw new Error('Service endpoint uri is required');
+    }
```

Comment:
I recommend checking `type` and `endpoint` against typeof 'string' like elsewhere... and doing this as a general pattern if doing parameter validation. There's some simple lib that @mattcollier likes to use for that purpose (can't remember the name) that keeps the code shorter ... but not sure if we should do that here unless it's client-side friendly.

Reply:
you're referring to https://www.npmjs.com/package/assert-plus which I don't think is browser friendly.